Adding view of DogStatsD metrics received by the agent into agent flare

DDuongNguyen opened the pull request:
### What does this PR do?
When `dogstatsd_metrics_stats_enable` is set to `true`, in addition to having the ability to run `agent dogstatsd_stats` for the stats of the metrics that are being collected by the agent from dogstatsd, it also automatically generate dogstatsd-stats log file that keep track of all metric instances that are sent to the agent. 
Example of what the [log file](https://a.cl.ly/llugeybd) look like:
![Image 2023-01-27 at 5 40 17 PM](https://user-images.githubusercontent.com/47346352/215219839-ef8c53bb-fe8a-4ff3-ac7c-f42b283a6bfc.jpg)
![Image 2023-01-27 at 5 42 06 PM](https://user-images.githubusercontent.com/47346352/215219944-af20f44e-7652-409c-82c9-04670af41149.jpg)

### Motivation
Challenges:
Currently, the support team lacks method to troubleshoot missing Dogstatsd metrics. 
Support generally will have to either recreate similar logic as the customer's code to test out the metric to identify if the missing metric is happening between DogstatsD and the Agent or the Agent and our endpoint. 

Solution:
This feature allows some logging on the agent so we can identify whether a Dogstatsd metric has been successfully delivered to the agent. 

The log file specifically will also help in identifying time bucket when a metric goes missing. 

### Additional Notes
This feature also introduce 2 new env var for the agent config:
- `dogstatsd_log_file`: file path with default path `datadog-agent/logs/dogstatsd_info/dogstatsd-stats.log`
- `dogstatsd_log_file_max_rolls` with default value being: `3` at 10mb

### Possible Drawbacks / Trade-offs

### Describe how to test/QA your changes
- `datadog.yaml` config includes:

```
api_key: <api_key>
use_dogstatsd: true
dogstatsd_port: 8125
dogstatsd_metrics_stats_enable: true
dogstatsd_log_file_max_rolls: 5
log_file_max_size: 3072
```

Logging is sucessful if dogstatsd log file are generated with path: `/dogstatsd_info/dogstatsd-stats.log*` by default along with 6 different log files 

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

scottopell commented on pkg/util/constants/constant_darwins.go:
```diff
@@ -0,0 +1,6 @@
+package constants
```
File name typo, should be `constant_darwin.go`, not `constant_darwins.go`.

Likely the cause of the lint failure below

DDuongNguyen replied:
Name changed

scottopell commented on pkg/util/constants/constants_nix.go:
```diff
@@ -0,0 +1,9 @@
+
+const (
+	//DefaultDogstatsDLogFile points to the dogstatsd stats log file that will be used if not configured
+	DefaultDogstatsDLogFile = "/var/log/datadog/dogstatsd_info/dogstatsd-stats.log"
```
I support the usage of this new `constants` pkg.
Can we move the constants from the existing `cmd/agent/common` pkg into this too?

DDuongNguyen replied:
I moved all cons and vars from `common` pkg to `constants` pkg. Also change all other codes that are using the cons from `common` to `constants` pkg accordingly

scottopell commented on pkg/config/config_template.yaml:
```diff
@@ -1738,6 +1738,20 @@ api_key:
 #
 # dogstatsd_stats_enable: false
 
+## @param dogstatsd_log_file - string - optional
```
I would lean towards leaving these two config options out of the default template, these feel like they'll be very infrequently used config options. 

It could be useful to add a note saying that the stats will be written to a log file in the above `dogstatsd_stats_enable` block though. Users may not want to have metric names/tags visible on disk, and it's not obvious that the `stats_enable` will now do that.

DDuongNguyen replied:
I removed the 2 env vars in the yaml file and added comment on how enabling the `dogstatsd_stats_enable` env var will now generate dogstatsd metric logs

scottopell commented on pkg/dogstatsd/server.go:
```diff
@@ -341,6 +345,41 @@ func NewServer(demultiplexer aggregator.Demultiplexer, serverless bool) (*Server
+		logFile = constants.DefaultDogstatsDLogFile
+	}
+
+	// Configuring max roll for log file, if dogstatsd_log_file_max_rolls env var is not set within datadog.yaml then default value is 3
```
nit: two separate ways of setting this option, one is env var, the other is within `datadog.yaml`. Can simplify this by saying "if dogstatsd_log_file_max_rolls config option is not set"

DDuongNguyen replied:
I changed the logic to:
```
if dogstatsd_log_file_max_rolls == 0 {
		config.Datadog.Set("dogstatsd_log_file_max_rolls", dogstatsd_log_file_max_rolls)
	}
```

scottopell commented on pkg/dogstatsd/server.go:
```diff
@@ -341,6 +345,41 @@ func NewServer(demultiplexer aggregator.Demultiplexer, serverless bool) (*Server
+		slog.Error(err)
+	}
+
+	// slog.UseLogger(logger)
```
Is this needed?

scottopell commented on pkg/dogstatsd/server.go:
```diff
@@ -782,6 +822,13 @@ func (s *Server) storeMetricStats(sample metrics.MetricSample) {
+	count := fmt.Sprintf(" Count: %d |", ms.Count)
+	ls := fmt.Sprintf(" Last Seen : %d", ms.LastSeen)
+
+	s.Logger.Info(name + tags + count + ls)
```
How frequently does this function run? Roughly how long will it take to fill up the 10mb log file?

DDuongNguyen replied:
- Brian and I talked and this function is in a hot path so it get ran everytime a [server worker](https://github.com/DataDog/datadog-agent/blob/main/pkg/dogstatsd/server_worker.go#L50-L66) is ran.

- I took a quick test sending roughly 40k metrics with 1s pause each round. This took a whole day to filled up the 10MB. So ideally if they turn on the env var, we would have roughly 3 days worth of metrics as rollover

scottopell commented on pkg/util/constants/constant_windows.go:
```diff
@@ -0,0 +1,6 @@
+package constants
+
+var (
+	//DefaultDogstatsDLogFile points to the dogstatsd stats log file that will be used if not configured
```
```suggestion
	// DefaultDogstatsDLogFile points to the dogstatsd stats log file that will be used if not configured
```
Same for the other constant comments

scottopell commented on pkg/dogstatsd/server.go:
```diff
@@ -171,6 +172,9 @@ type Server struct {
 
 	enrichConfig enrichConfig
+
+	// logger is an instant of the logger config that can be used to create new logger for dogstatsd-stats metrics
```
```suggestion
	// logger is an instance of the logger config that can be used to create new logger for dogstatsd-stats metrics
```

scottopell commented on pkg/dogstatsd/server.go:
```diff
@@ -171,6 +172,9 @@ type Server struct {
 	enrichConfig enrichConfig
+
+	// logger is an instant of the logger config that can be used to create new logger for dogstatsd-stats metrics
+	Logger slog.LoggerInterface
```
```suggestion
	StatsLogger slog.LoggerInterface
```

DDuongNguyen replied:
I changed it to `DogstatsdDebugLogger` so it would be easier to understand in the future

scottopell commented on pkg/dogstatsd/server.go:
```diff
@@ -341,6 +345,41 @@ func NewServer(demultiplexer aggregator.Demultiplexer, serverless bool) (*Server
+		"info",
+		"common",
+		"",
+		"%n %Date %Time | [DOGSTATSD_STATSD] | [%LEV] %Msg %n",
```
```suggestion
		"%n %Date %Time | [DOGSTATSD_STATS] | [%LEV] %Msg %n",
```

gh123man replied:
Also can remove the leading `%n` as that is adding a blank line between every log line in the file. 
[Format docs are here for reference ](https://github.com/cihub/seelog/wiki/Format-reference)

gh123man commented on pkg/dogstatsd/seelog_config.go:
```diff
@@ -0,0 +1,99 @@
+}
+
+const seelogConfigurationTemplate = `
+<seelog minlevel="{{.logLevel}}">
```
This file is mostly Copy/Pasted from https://github.com/DataDog/datadog-agent/blob/main/pkg/config/internal/seelog/seelog_config.go

I think we can improve this by either: 
1. making `seelog/seelog_config.go` non-internal (since we now want to use it in more than one place)
2. Or - trim this file down to just the necessary bits. In fact, you probably don't need the majority of what is defined here and can just inline it with your implementation (where you call `NewSeelogConfig` and `Render`)

DDuongNguyen replied:
I opted into moving the `seelog_config.go` outside of the internal folder and changed other files import that were using the internal file

gh123man commented on pkg/dogstatsd/server.go:
```diff
@@ -376,6 +415,7 @@ func NewServer(demultiplexer aggregator.Demultiplexer, serverless bool) (*Server
 			serverlessMode:            serverless,
 			originOptOutEnabled:       config.Datadog.GetBool("dogstatsd_origin_optout_enabled"),
 		},
+		Logger: logger,
```
Lets call this `DogstatsdDebugLogger` (or similar) so a future reader knows what its for.

gh123man commented on cmd/agent/subcommands/run/command.go:
```diff
@@ -57,11 +53,15 @@ import (
 	"github.com/DataDog/datadog-agent/pkg/util/hostname"
 	pkglog "github.com/DataDog/datadog-agent/pkg/util/log"
 	"github.com/DataDog/datadog-agent/pkg/version"
+	"github.com/spf13/cobra"
```
NIT: it seems like these imports are getting moved from the top section to the bottom here. Can you move them back (here and in other files)? (VSCode + the go plugin *should* respect this)

We generally keep `DataDog` level imports in their own group.

DDuongNguyen replied:
done ✅

gh123man commented:
CI is failing. 
[You are missing an import to satisfy this line](https://github.com/DataDog/datadog-agent/blob/duong.nguyen/dogstatsd-stats-log/cmd/agent/common/helpers.go#L62)

[You have an unused import](https://github.com/DataDog/datadog-agent/blob/duong.nguyen/dogstatsd-stats-log/cmd/agent/subcommands/jmx/command_test.go#L18)


You can run the tests locally with `inv test --skip-linters` and it will show you these problems.

gh123man commented on pkg/dogstatsd/server.go:
```diff
@@ -171,6 +174,9 @@ type Server struct {
 
 	enrichConfig enrichConfig
+
+	// logger is an instance of the logger config that can be used to create new logger for dogstatsd-stats metrics
```
```suggestion
	// DogstatsdDebugLogger is an instance of the logger config that can be used to create new logger for dogstatsd-stats metrics
```

gh123man commented on pkg/dogstatsd/server.go:
```diff
@@ -341,6 +347,39 @@ func NewServer(demultiplexer aggregator.Demultiplexer, serverless bool) (*Server
+	}
+	logger, err := slog.LoggerFromConfigAsString(seelogConfigStr)
+	if err != nil {
+		slog.Error(err)
```
This can be a `log.Errorf` instead of `slog.Error` (same for line `376`)

gh123man reviewed (changes requested):
Just a couple more small comments, I think we are almost there.

gh123man commented:
Also - I think we can add a changelog (reno) to this PR.

NouemanKHAL reviewed (approved):
Approving for the files owned by @platform-integrations: 
- `cmd/cluster-agent-cloudfoundry/app/app.go`
- `cmd/cluster-agent-cloudfoundry/app/flare.go`

vickenty commented on pkg/dogstatsd/server.go:
```diff
@@ -341,6 +347,39 @@ func NewServer(demultiplexer aggregator.Demultiplexer, serverless bool) (*Server
+
+	// Configuring max roll for log file, if dogstatsd_log_file_max_rolls env var is not set within datadog.yaml then default value is 3
+	dogstatsd_log_file_max_rolls := config.Datadog.GetInt64("dogstatsd_log_file_max_rolls")
+	if dogstatsd_log_file_max_rolls == 0 {
```
What if `dogstatsd_log_file_max_rolls` is less than zero?

Would it make sense to emit a log warning if the value is invalid? I think we do so in other places.

vickenty commented on pkg/dogstatsd/server.go:
```diff
@@ -782,6 +822,13 @@ func (s *Server) storeMetricStats(sample metrics.MetricSample) {
+	count := fmt.Sprintf(" Count: %d |", ms.Count)
+	ls := fmt.Sprintln(" Last Seen : ", ms.LastSeen)
+
+	s.DogstatsdDebugLogger.Info(name + tags + count + ls)
```
Would it make sense to use `Infof` here?

vickenty commented on pkg/dogstatsd/server.go:
```diff
@@ -341,6 +347,39 @@ func NewServer(demultiplexer aggregator.Demultiplexer, serverless bool) (*Server
+	if err != nil {
+		log.Error(err)
+	}
+	logger, err := slog.LoggerFromConfigAsString(seelogConfigStr)
```
What `slog.LoggerFromConfigAsString` returns on error? Do we want to use the `logger` variable if there was an error?

DDuongNguyen replied:
I changed the logic a bit so now if `slog.LoggerFromConfigAsString` errors out, the agent logger will log the errors

vickenty commented on pkg/dogstatsd/server.go:
```diff
@@ -341,6 +347,39 @@ func NewServer(demultiplexer aggregator.Demultiplexer, serverless bool) (*Server
+
+	seelogConfigStr, err := cfg.Render()
+	if err != nil {
+		log.Error(err)
```
What happens if configuration fails to render? Is it still safe to pass to `LoggerFromConfigAsString`?

DDuongNguyen replied:
if the config fails then the default behavior would be logging to console but it won't save to any log file so i guess minimal performance impact will be the result

DDuongNguyen replied:
+ I also added a null check when the the log is supposed to to be used in case the error cause a panic error.

vickenty commented on pkg/config/config_template.yaml:
```diff
@@ -1735,6 +1735,7 @@ api_key:
 ## @param dogstatsd_stats_enable - boolean - optional - default: false
 ## @env DD_DOGSTATSD_STATS_ENABLE - boolean - optional - default: false
 ## Publish DogStatsD's internal stats as Go expvars.
+## Enabling this will write DogstatsD metrics that agent received to log files
```
Is there a way to disable this log file, but keep stats enabled?

It may be better to have a separate option for this feature.

scottopell replied:
This could be a good motivation for adding a separate config `dogstatsd_log_file_max_size`, have it default to the same as the agent log and if its set to zero that would disable the `dogstatsd_stats` logging. What do you think?

DDuongNguyen replied:
So just checking the logic here, but if that is the case, i think it makes more sense to just change the env var name all together to something more like:
```
dogstatsd_stats_logging: 0 or 1 
```
with 0 meaning disabled dogstatsd stats logging
and 1 for using the agent's `log_file_max_rolls` to determine how many log file will be rolled over. 

Is that correct?

DDuongNguyen replied:
+ I added the suggested env var so now there are 3 conditions for a log to be generated on the log file:
1 and 2 being these 2 env vars:
```
dogstatsd_metrics_stats_enable: true
dogstatsd_stats_logging: false 
```
and 3rd condition being a null check for the logger.

vickenty commented on pkg/dogstatsd/server.go:
```diff
@@ -782,6 +822,13 @@ func (s *Server) storeMetricStats(sample metrics.MetricSample) {
+	name := fmt.Sprintf("Metric Name: %s |", ms.Name)
+	tags := fmt.Sprintf(" Tags: {%s} |", ms.Tags)
+	count := fmt.Sprintf(" Count: %d |", ms.Count)
+	ls := fmt.Sprintln(" Last Seen : ", ms.LastSeen)
```
Would it make sense to use a human-readable representation of the timestamp instead of the raw view?

DDuongNguyen replied:
I updated the format to more like this:
```2023-02-13 14:13:21 | [DOGSTATSD_STATS] | [INF] Metric Name: dd_mac_book_count_metric.count | Tags: {environment:prod  dogstatsd_metric:true  secondary_tag:new_metric?} | Count: 1 | Last Seen: 2023-02-13 14:13:03.657791 -0500 EST m=+3.196348841```

vickenty commented on pkg/dogstatsd/server.go:
```diff
@@ -341,6 +347,39 @@ func NewServer(demultiplexer aggregator.Demultiplexer, serverless bool) (*Server
+	}
+
+	// Configure the logger
+	cfg.EnableFileLogging(logFile, config.Datadog.GetSizeInBytes("log_file_max_size"), uint(config.Datadog.GetInt("dogstatsd_log_file_max_rolls")))
```
Would it make sense to have a separate tunable for the log file size?

DDuongNguyen replied:
I'll add another env var for log file size as:
```dogstats_log_file_max_size```

vickenty commented on pkg/dogstatsd/server.go:
```diff
@@ -341,6 +347,39 @@ func NewServer(demultiplexer aggregator.Demultiplexer, serverless bool) (*Server
+		"info",
+		"common",
+		"",
+		"%Date %Time | [DOGSTATSD_STATS] | [%LEV] %Msg ",
```
Since the log file has only one type of entries, do we really need `| [DOGSTATSD_STATS] | [%LEV]` part?

DDuongNguyen replied:
This might be for more future proofing but I guess for now its not needed, i'll remove it instead then

vickenty commented on pkg/dogstatsd/server.go:
```diff
@@ -341,6 +347,39 @@ func NewServer(demultiplexer aggregator.Demultiplexer, serverless bool) (*Server
+	// Configuring the log file path
+	logFile := config.Datadog.GetString("dogstatsd_log_file")
+	if logFile == "" {
+		logFile = constants.DefaultDogstatsDLogFile
```
Could this be the default value in `BindEnvAndSetDefault`?

scottopell replied:
I don't think it could be since we have different defaults for different platforms

vickenty commented on pkg/dogstatsd/server.go:
```diff
@@ -341,6 +347,39 @@ func NewServer(demultiplexer aggregator.Demultiplexer, serverless bool) (*Server
+		false)
+
+	// Configuring the log file path
+	logFile := config.Datadog.GetString("dogstatsd_log_file")
```
I wonder what happens if this is set to the path to the main `agent.log` file?

scottopell replied:
Its a good thought, but we don't handle this mistake in the `jmx_log_file` for example, so maybe this kind of extra error-checking would be best left for a future PR. (would probably be a good "first-change" task for a new embed/intern)

scottopell commented on cmd/cluster-agent-cloudfoundry/app/flare.go:
```diff
@@ -0,0 +1,131 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
```
This file appears shouldn't be here, I imagine this was a git merge gone wrong. @DDuongNguyen , can you `git rm` this file and `cmd/cluster-agent-cloudfoundry/subcommands/run/command.go`? 

We'll squash-merge this branch, so no need to try to fix up the old merge or anything like that.

gh123man commented on cmd/cluster-agent-cloudfoundry/subcommands/run/command.go:
```diff
@@ -1,253 +0,0 @@
-
-package run
-
-import (
```
This file was deleted by mistake and is causing CI to fail.

scottopell commented:
I don't fully understand the reno failure here. I can replicate locally and `reno` is claiming that the same file is present multiple times in the same git commit, but I only see it once in the directory tree.
```
$ reno lint
...
ValueError: dd29f9ae1c40e18d: found several files in one commit (b'04b7430ded26718d579990a13685b3f23fc40474') with the same UID: ['dogstatsd-adding-dogstats-log-file-in-flare-dd29f9ae1c40e18d.yaml', 'dogstatsd-adding-dogstats-log-file-in-flare-dd29f9ae1c40e18d.yaml']
$ git co 04b7430ded26718d579990a13685b3f23fc40474
$ find . -name "*dd29f9ae1c40e18d.yaml" | wc -l
1
```

The commit it flags says in the commit msg there were local changes (adding release note) and a merge in the same commit, so my guess is that this is confusing to `reno` in some way, but my git/reno knowledge starts to break down past this :) 

My best suggestion to get unblocked would be to squash this branch into a single commit and force-push and I imagine this check will start passing.

kayayarai commented on pkg/config/config_template.yaml:
```diff
@@ -1738,6 +1738,17 @@ api_key:
 # dogstatsd_stats_enable: false
 
+## @param dogstatsd_stats_logging - boolean - optional - default: true
+## Set to true to write DogstatsD metrics that the agent received to dogstats_stats log files
```
```suggestion
## Set to true to write DogstatsD metrics received by the Agent to dogstats_stats log files.
```

kayayarai commented on pkg/config/config_template.yaml:
```diff
@@ -1738,6 +1738,17 @@ api_key:
+#
+# dogstatsd_stats_logging: true
+
+## @param dogstatsd_log_file_max_size - custom - optional - default: 10MB
+## Maximum size of dogstatsd log file. Use either a size (e.g. 10MB) or
+## provide value in bytes: 10485760
```
```suggestion
## @param dogstatsd_log_file_max_size - custom - optional - default: 10MB
## Maximum size of dogstatsd log file. Use either a size (for example, 10MB) or
## provide value in bytes (for example, 10485760).
```

kayayarai commented on releasenotes/notes/dogstats-stats-logging-b7e38c747ee8d8c7.yaml:
```diff
@@ -0,0 +1,12 @@
+---
+features:
+  - |
+    Enabling ``dogstatsd_metrics_stats_enable`` will now enable ``dogstatsd_stats_logging``. ``dogstatsd_stats_logging`` when abled will generate dogstatsd log files at:
```
```suggestion
    Enabling ``dogstatsd_metrics_stats_enable`` will now enable ``dogstatsd_stats_logging``. When enabled, ``dogstatsd_stats_logging`` generates dogstatsd log files at:
```

kayayarai commented on releasenotes/notes/dogstats-stats-logging-b7e38c747ee8d8c7.yaml:
```diff
@@ -0,0 +1,12 @@
+      - For ``Windows``: ``c:\programdata\datadog\logs\dogstatsd_info\dogstatsd-stats.log``
+      - For ``Linux``: ``/var/log/datadog/dogstatsd_info/dogstatsd-stats.log``
+      - For ``MacOS``: ``/opt/datadog-agent/logs/dogstatsd_info/dogstatsd-stats.log``
+    These log files will also be automatically attached to the flare  
```
```suggestion
    These log files are also automatically attached to the flare.
```

kayayarai commented on releasenotes/notes/dogstats-stats-logging-b7e38c747ee8d8c7.yaml:
```diff
@@ -0,0 +1,12 @@
+      - For ``MacOS``: ``/opt/datadog-agent/logs/dogstatsd_info/dogstatsd-stats.log``
+    These log files will also be automatically attached to the flare  
+  - |
+    Users can also adjust the dogstatsd-stats logging by using:
```
```suggestion
    You can adjust the dogstatsd-stats logging by using:
```

kayayarai reviewed (approved):
some edit suggestions, and 👍

DDuongNguyen committed:
squash merging

DDuongNguyen committed:
+ Language edit suggestions.

DDuongNguyen committed:
adjusting constants path for windows

clarkb7 commented on cmd/agent/common/common_windows.go:
```diff
@@ -25,29 +25,14 @@ var (
-		DefaultConfPath = pd
-		DefaultLogFile = filepath.Join(pd, "logs", "agent.log")
-		DefaultDCALogFile = filepath.Join(pd, "logs", "cluster-agent.log")
+		constants.DefaultConfPath = pd
```
I'm worried that with the constants moved to another package that this package might not be imported and the constants won't be updated. Should an `init` be added to `constants_windows.go` and these path updates moved there? It looks like some of the constants aren't updated, too.

DDuongNguyen replied:
That makes sense 👍 I'll adjust it 🙇

bkabrda reviewed (approved):
LGTM for agent-platform :+1:

DDuongNguyen committed:
switching init logic from common_windows to constants_windows

DDuongNguyen committed:
adjusting pkgs

vickenty commented on releasenotes/notes/dogstatsd-stats-logging-b7e38c747ee8d8c7.yaml:
```diff
@@ -0,0 +1,12 @@
+---
+features:
+  - |
+    Enabling ``dogstatsd_metrics_stats_enable`` will now enable ``dogstatsd_stats_logging``. ``dogstatsd_stats_logging``.  When enabled, ``dogstatsd_stats_logging`` generates dogstatsd log files at:
```
Setting name repeats twice:
```
``dogstatsd_stats_logging``. ``dogstatsd_stats_logging``.
```

vickenty commented on pkg/config/config.go:
```diff
@@ -248,10 +248,14 @@ func InitConfig(config Config) {
 	config.BindEnvAndSetDefault("confd_path", defaultConfdPath)
 	config.BindEnvAndSetDefault("additional_checksd", defaultAdditionalChecksPath)
 	config.BindEnvAndSetDefault("jmx_log_file", "")
+	config.BindEnvAndSetDefault("dogstatsd_log_file", "")
 	config.BindEnvAndSetDefault("log_payloads", false)
 	config.BindEnvAndSetDefault("log_file", "")
 	config.BindEnvAndSetDefault("log_file_max_size", "10Mb")
 	config.BindEnvAndSetDefault("log_file_max_rolls", 1)
+	config.BindEnvAndSetDefault("dogstatsd_stats_logging", true)
+	config.BindEnvAndSetDefault("dogstatsd_log_file_max_rolls", 3)
+	config.BindEnvAndSetDefault("dogstatsd_log_file_max_size", "10Mb")
```
Would it make sense to keep these settings together with the rest of dogstatsd options? (Somewhere around https://github.com/datadog/datadog-agent/blob/563ac9ff40cb5755a71728f4bd768a738d6c8f97/pkg/config/config.go#L523)

vickenty commented on pkg/dogstatsd/server.go:
```diff
@@ -289,6 +298,42 @@ func NewServer(serverless bool) *Server {
+	}
+
+	// Configure the logger
+	cfg.EnableFileLogging(logFile, config.Datadog.GetSizeInBytes("dogstatsd_log_file_max_size"), uint(dogstatsd_log_file_max_rolls))
```
Looking at `SetupJMXLogger` in `config/log.go`, it seems to be busy with the much the same task. Would it make sense to move this code to `config/log.go`? It would allow to keep `seelog_config` internal to config package.

sgnn7 commented on pkg/util/constants/constant_windows.go:
```diff
@@ -0,0 +1,41 @@
+	"github.com/DataDog/datadog-agent/pkg/util/winutil"
+)
+
+package constants
```
Linter is correct here - package should go above imports

sgnn7 commented on pkg/config/config_template.yaml:
```diff
@@ -1738,6 +1738,17 @@ api_key:
+## @param dogstatsd_stats_logging - boolean - optional - default: true
+## Set to true to write DogstatsD metrics received by the Agent to dogstats_stats log files.
+#
+# dogstatsd_stats_logging: true
```
Should this be `dogstatsd_logging_enabled` or something similar?

DDuongNguyen replied:
👍

sgnn7 commented on pkg/util/constants/constant_darwin.go:
```diff
@@ -0,0 +1,23 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
```
I'm not sure that `constants` nor the full import path is descriptive enough name to tell what is contained within it. What about `pkg/common/paths/*` or similar?

NouemanKHAL reviewed (approved):
Looks good for @DataDog/platform-integrations

DDuongNguyen commented on pkg/dogstatsd/server.go:
```diff
@@ -289,6 +298,42 @@ func NewServer(serverless bool) *Server {
+
+	// Configuring max roll for log file, if dogstatsd_log_file_max_rolls env var is not set (or set improperly ) within datadog.yaml then default value is 3
+	dogstatsd_log_file_max_rolls := config.Datadog.GetInt("dogstatsd_log_file_max_rolls")
+	if dogstatsd_log_file_max_rolls < 1 {
```
I guess this goes down to whether we use our logic or seelog logic, only down side of using seelog logic i think is that if they change this specific logic in the future then our code might not work properly because of that edge case.

vickenty replied:
How many files there will be on disk the agent runs with max_rolls=1?

DDuongNguyen replied:
Changed logic to < 0

DDuongNguyen commented on pkg/dogstatsd/server.go:
```diff
@@ -325,6 +370,8 @@ func NewServer(serverless bool) *Server {
 			serverlessMode:            serverless,
 			originOptOutEnabled:       config.Datadog.GetBool("dogstatsd_origin_optout_enabled"),
 		},
+		DogstatsdLogging:     config.Datadog.GetBool("dogstatsd_stats_logging"),
+		DogstatsdDebugLogger: dogstatsdLogger,
```
changed it to private.

DDuongNguyen committed:
+ Moved dogstatsd logger set up logic to log.go
+ Add condition in config_template that dogstatsd_stats_enable is required to be enabled for logging to generate
+ Adjust test file so the server_test.go file does not generate log when testing dogstatsd metric stats
+ Change naming from dogstatsd_stats_logging to dogstatsd_logging_enabled

DDuongNguyen committed:
Merge remote-tracking branch 'origin/main' into duong.nguyen/dogstatsd-stats-log
